File: keyway/src/domains/access/entity/access.rs
//! How far a caller gets on one secret.
//!
//! This is the whole authorisation test, and it is deliberately small. Under
//! ADR-0002 a delegation carries its own level and nothing caps it, so there
//! is no ceiling to intersect and no role to consult: the answer is ownership,
//! or the grant addressed to this caller, or nothing.
//!
//! Keeping it in one function is the point. "Who can see this secret, and how
//! far" has to be answerable by reading one list, and that is only true if
//! every code path asks the same question the same way.

use super::{Delegation, Level, Ownership};
use crate::domains::identity::entity::Actor;
use chrono::{DateTime, Utc};

/// What a caller may do with one secret, and why.
///
/// The reason is carried because a refusal owes its reader a sentence, and
/// because the audit log records the basis on which a reveal was allowed.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Access {
    pub level: Option<Level>,
    pub basis: Basis,
    /// Which keys of a key/value secret are open, or `None` for all of them.
    pub keys: Option<Vec<String>>,
}

/// Why a caller may do what they may.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Basis {
    /// Nothing opens this secret for this caller.
    Nothing,
    /// They own it, so they run it outright whatever role they hold.
    Owner,
    /// The operational bypass.
    Admin,
    /// A grant, and who it was addressed to — a handle or a group name, which
    /// is what a person needs in order to know where their access came from.
    Delegated { subject: String },
}

impl Access {
    /// Nothing at all.
    #[must_use]
    pub fn none() -> Self {
        Self {
            level: None,
            basis: Basis::Nothing,
            keys: None,
        }
    }

    /// Whether this opens the secret at least as far as `wanted`.
    #[must_use]
    pub fn allows(&self, wanted: Level) -> bool {
        self.level.is_some_and(|held| held >= wanted)
    }

    /// Whether this opens `key` of a key/value secret at `wanted`.
    #[must_use]
    pub fn allows_key(&self, wanted: Level, key: &str) -> bool {
        self.allows(wanted)
            && self
                .keys
                .as_ref()
                .is_none_or(|keys| keys.iter().any(|k| k == key))
    }

    /// Whether the secret's existence should be disclosed at all.
    #[must_use]
    pub fn is_visible(&self) -> bool {
        self.level.is_some()
    }
}

/// How far `actor` gets on the secret that `ownership` and `grants` describe.
///
/// `grants` is every delegation written against that secret, whoever it is
/// addressed to — this function picks out the ones that name the caller. That
/// is deliberate: a caller-filtered query would put half the authorisation
/// rule in SQL, where the next person to write a query is free to get it
/// wrong.
#[must_use]
pub fn resolve(
    actor: &Actor,
    ownership: Option<&Ownership>,
    grants: &[Delegation],
    now: DateTime<Utc>,
) -> Access {
    // An owner runs their secret outright: change it, delegate it, revoke a
    // grant, transfer it. Checked before admin so the audit basis names the
    // narrower, truer reason.
    if ownership.is_some_and(|owned| owned.owner == actor.handle()) {
        return Access {
            level: Some(Level::Write),
            basis: Basis::Owner,
            keys: None,
        };
    }

    if actor.is_admin() {
        return Access {
            level: Some(Level::Write),
            basis: Basis::Admin,
            keys: None,
        };
    }

    // The best active grant addressed to this caller. Best rather than first:
    // a person may be named directly and through a group, and the answer is
    // the most that was granted, not whichever row the database returned
    // first.
    let mut best: Option<&Delegation> = None;
    for grant in grants {
        if !grant.is_active(now) || !actor.is_addressed_by(&grant.subject) {
            continue;
        }
        if best.is_none_or(|held| grant.level > held.level) {
            best = Some(grant);
        }
    }

    best.map_or_else(Access::none, |grant| Access {
        level: Some(grant.level),
        basis: Basis::Delegated {
            subject: grant.subject.id().to_owned(),
        },
        keys: grant.scoped_keys().map(<[String]>::to_vec),
    })
}

#[cfg(test)]
#[path = "access_tests.rs"]
mod tests;

File: internal/domains/access/entity/access.go
// Package entity is how far a caller gets on one secret.
//
// This is the whole authorisation test, and it is deliberately small. Under
// ADR-0002 a delegation carries its own level and nothing caps it, so there
// is no ceiling to intersect and no role to consult: the answer is ownership,
// or the grant addressed to this caller, or nothing.
//
// Keeping it in one function is the point. "Who can see this secret, and how
// far" has to be answerable by reading one list, and that is only true if
// every code path asks the same question the same way.
package entity

import "time"

// Caller is who is asking, as this domain needs to see them.
//
// In Rust this was the identity domain's Actor outright; in Go that import
// would be a cycle, because the Actor answers in this package's Subject and
// Level. The interface carries exactly the questions access asks, and the
// identity Actor satisfies it.
type Caller interface {
	// Handle is the name every service keys and logs on.
	Handle() string
	// IsAdmin is whether this caller holds the operational bypass.
	IsAdmin() bool
	// IsAddressedBy is whether a delegation addressed to `subject` is
	// addressed to THIS caller.
	IsAddressedBy(subject Subject) bool
	// Subjects is every string a delegation could name this caller by.
	Subjects() []Subject
}

// Access is what a caller may do with one secret, and why.
//
// The reason is carried because a refusal owes its reader a sentence, and
// because the audit log records the basis on which a reveal was allowed.
type Access struct {
	// Level is nil when nothing opens the secret at all.
	Level *Level
	Basis Basis
	// Keys is which keys of a key/value secret are open, or nil for all of
	// them.
	Keys []string
}

// Basis is why a caller may do what they may.
//
// A comparable value rather than a set of types, so a test and an audit line
// can ask "is this exactly Owner" with ==.
type Basis struct {
	kind string
	// subject is who a delegated grant was addressed to — a handle or a group
	// name, which is what a person needs in order to know where their access
	// came from.
	subject string
}

// BasisNothing: nothing opens this secret for this caller.
var BasisNothing = Basis{kind: "nothing"}

// BasisOwner: they own it, so they run it outright whatever role they hold.
var BasisOwner = Basis{kind: "owner"}

// BasisAdmin: the operational bypass.
var BasisAdmin = Basis{kind: "admin"}

// BasisDelegated: a grant, and who it was addressed to.
func BasisDelegated(subject string) Basis {
	return Basis{kind: "delegated", subject: subject}
}

// DelegatedTo is who the grant behind this basis was addressed to, when it
// was a grant at all.
func (b Basis) DelegatedTo() (string, bool) {
	return b.subject, b.kind == "delegated"
}

func (b Basis) String() string {
	if b.kind == "delegated" {
		return "delegated:" + b.subject
	}
	return b.kind
}

// NoAccess is nothing at all.
func NoAccess() Access {
	return Access{Level: nil, Basis: BasisNothing, Keys: nil}
}

// Allows is whether this opens the secret at least as far as `wanted`.
func (a Access) Allows(wanted Level) bool {
	return a.Level != nil && *a.Level >= wanted
}

// AllowsKey is whether this opens `key` of a key/value secret at `wanted`.
func (a Access) AllowsKey(wanted Level, key string) bool {
	if !a.Allows(wanted) {
		return false
	}
	if a.Keys == nil {
		return true
	}
	for _, k := range a.Keys {
		if k == key {
			return true
		}
	}
	return false
}

// IsVisible is whether the secret's existence should be disclosed at all.
func (a Access) IsVisible() bool {
	return a.Level != nil
}

// Resolve is how far `actor` gets on the secret that `ownership` and `grants`
// describe.
//
// `grants` is every delegation written against that secret, whoever it is
// addressed to — this function picks out the ones that name the caller. That
// is deliberate: a caller-filtered query would put half the authorisation
// rule in SQL, where the next person to write a query is free to get it
// wrong.
func Resolve(actor Caller, ownership *Ownership, grants []Delegation, now time.Time) Access {
	// An owner runs their secret outright: change it, delegate it, revoke a
	// grant, transfer it. Checked before admin so the audit basis names the
	// narrower, truer reason.
	if ownership != nil && ownership.Owner == actor.Handle() {
		return Access{Level: levelOf(Write), Basis: BasisOwner, Keys: nil}
	}

	if actor.IsAdmin() {
		return Access{Level: levelOf(Write), Basis: BasisAdmin, Keys: nil}
	}

	// The best active grant addressed to this caller. Best rather than first:
	// a person may be named directly and through a group, and the answer is
	// the most that was granted, not whichever row the database returned
	// first.
	var best *Delegation
	for i := range grants {
		grant := &grants[i]
		if !grant.IsActive(now) || !actor.IsAddressedBy(grant.Subject) {
			continue
		}
		if best == nil || grant.Level > best.Level {
			best = grant
		}
	}

	if best == nil {
		return NoAccess()
	}
	return Access{
		Level: levelOf(best.Level),
		Basis: BasisDelegated(best.Subject.ID()),
		Keys:  best.ScopedKeys(),
	}
}

func levelOf(l Level) *Level {
	return &l
}
